docs/enhancements/dynamic-fetching-of-openshift-versions.md

```diff
@@ -0,0 +1,301 @@
+
+The idea of this enhancement is configuring the assisted-service to the OpenShift
+channels it needs to follow (e.g. ``candidate-4.12``, ``stable-4.11``), and let
+it periodically fetch versions to provide them to users.
+
+## Motivation
+
+It becomes a pretty tedious process to handle each version separately. Some
+use-cases we tend to have:
+* A certain user uses a specific version x.y.z for certification, and we replace
+  this version with a newer one x.y.z+w (w>=1). In this case we'll append again the
+  x.y.z version to the configuration.
+* As described above, we do version updates either automatically or manually.
+* QE team wants to check if a newly added OpenShift version has a regression in its
+  behavior from a previous version.
+* QE team tests a flow which includes OpenShift upgrade process.
+
+### Goals
+
+* Assisted-service is able to provide all versions included in the upgrade-channels,
+  based on provided configuration.
+* We keep ourselves the option to specifically configure certain versions as we can
+  nowadays. This is to make sure we can handle edge-cases and be flexible if needed.
+* UI needs to be adjusted to handle this growing list of versions (at the time being,
+  a total of ~140 versions).
+
+### Non-Goals
+
+* Dynamic handling of y-versions, live-iso discovery images, and such. A new minor
+  OpenShift version is introduced every 4 months, requires additional code changes
+  (for example in feature-support-level list) and it isn't worth the costs and
+  implications of making assisted-image-service refresh itself and pull-in new
+  images.
+* Handling OpenShift versions in a kubernetes API interface. For this use-case
+  versions are externally configured in the ``ClusterImageSet`` custom resource.
+
+## Proposal
+
+The assisted-service will make use of a new configuration variable called
+``DYNAMIC_OPENSHIFT_RELEASES_CONFIG``, which outlines the supported matrix. For
+example:
+
+```
+DYNAMIC_OPENSHIFT_RELEASES_CONFIG='{
+   "4.11": {
+      "upgrade_channels": [
+         {
+            "cpu_architecture": "x86_64",
+            "channels": ["stable"]
+         }
+      ]
+   },
+   "4.12": {
+      "upgrade_channels": [
+         {
+            "cpu_architecture": "x86_64",
+            "channels": ["candidate", "stable"]
+         },
+         {
+            "cpu_architecture": "arm64",
+            "channels": ["candidate", "stable"]
+         },
+         {
+            "cpu_architecture": "multi",
+            "channels": ["candidate", "stable"]
+         }
+      ]
+   }
+}'
+```
+
+The assisted-service periodically polls versions from the ``upgrades_info`` API
+in the form of ``https://api.openshift.com/api/upgrades_info/v1/graph?channel=<channel>&arch=<arch>``.
+For example:
+```
+$ curl -s 'https://api.openshift.com/api/upgrades_info/v1/graph?channel=stable-4.11&arch=amd64' \
+    | jq -r '.nodes[] | select(.version | startswith("4.11")) | .version'
```

Comment:
An example that includes the payload may be more useful here as that's what we're really interested in.
Something like:

```
$ curl -s 'https://api.openshift.com/api/upgrades_info/v1/graph?channel=stable-4.11&arch=amd64' | jq -r '.nodes[] | .version + " => " + .payload'
4.10.9 => quay.io/openshift-release-dev/ocp-release@sha256:39f360002b9b5c730d1167879ad6437352d51e72acc9fe80add3ec2a0d20400d
4.10.35 => quay.io/openshift-release-dev/ocp-release@sha256:766bc6b381163ecccfea00556dd336dd49067413f69cbc29337fea778295c7bb
4.10.23 => quay.io/openshift-release-dev/ocp-release@sha256:e40e49d722cb36a95fa1c03002942b967ccbd7d68de10e003f0baa69abad457b
4.11.4 => quay.io/openshift-release-dev/ocp-release@sha256:e04ee7eea9e6a6764df218bcce3ee64e1816445158a0763954b599e0fc96580b
4.11.8 => quay.io/openshift-release-dev/ocp-release@sha256:6499bc69a0707fcad481c3cb73226c364586761c45b867dc31b345c6e6204e28
```

Reply:
done!